crates/worker/src/partition/state_machine/command_interpreter/mod.rs

```diff
@@ -1358,6 +1362,189 @@ where
+                if !is_completed {
+                    let_assert!(
+                        Entry::CompletePromise(CompletePromiseEntry {
+                            key,
+                            completion,
+                            ..
+                        }) = journal_entry.deserialize_entry_ref::<Codec>()?
+                    );
+
+                    if let Some(service_id) =
+                        invocation_metadata.invocation_target.as_keyed_service_id()
+                    {
+                        // Load state and write completion
+                        let promise_metadata = state.get_promise(&service_id, &key).await?;
+
+                        let completion_result = match promise_metadata {
+                            None => {
+                                // Just register the promise completion
+                                effects.put_promise(
+                                    service_id,
+                                    key,
+                                    PromiseMetadata {
+                                        state: PromiseState::Completed(completion),
+                                    },
+                                );
+                                CompletionResult::Empty
+                            }
+                            Some(PromiseMetadata {
+                                state: PromiseState::NotCompleted(listeners),
+                            }) => {
+                                // Send response to listeners
+                                for listener in listeners {
+                                    self.handle_outgoing_message(
+                                        OutboxMessage::ServiceResponse(InvocationResponse {
+                                            id: listener.invocation_id(),
+                                            entry_index: listener.journal_index(),
+                                            result: completion.clone().into(),
+                                        }),
+                                        effects,
+                                    );
+                                }
+
+                                // Now register the promise completion
+                                effects.put_promise(
+                                    service_id,
+                                    key,
+                                    PromiseMetadata {
+                                        state: PromiseState::Completed(completion),
+                                    },
+                                );
+                                CompletionResult::Empty
+                            }
+                            Some(PromiseMetadata {
+                                state: PromiseState::Completed(_),
+                            }) => {
+                                // Conflict!
+                                (&CONFLICT_INVOCATION_ERROR).into()
```

Comment:
Out of curiosity: How will this error surface to the user? Will it be a terminal error or returned as false on the complete call?

Reply:
For now it seems reasonable to return a terminal error. There might be other situations where we could fail to complete the promise that were discussed as well, e.g. fail completing a promise if the workflow didn't started yet.